src/renderers/WebGLRenderer.js

```diff
@@ -1982,9 +1982,7 @@ function WebGLRenderer( parameters ) {
 
 			WebGLUniforms.upload( _gl, materialProperties.uniformsList, m_uniforms, textures );
 
-		}
-
-		if ( material.isShaderMaterial && material.uniformsNeedUpdate === true ) {
+		} else if ( material.isShaderMaterial && material.uniformsNeedUpdate === true ) {
```

Comment:
Should `uniformsNeedUpdate` be set to false in the above `refreshMaterial` condition block, as well? If `refreshMaterial` _and_ `uniformsNeedUpdate` are true on the same frame then the uniforms will be updated but `uniformsNeedUpdate` will not be set to false meaning the uniforms will be updated unnecessarily the next frame, right?

Reply:
Um, but it's an `else if` statement...

Reply:
Right but consider this situation (or something similar):

**Frame 1**

- ShaderMaterial is newly created, uniforms set, and `uniformsNeedUpdate` set to true.
- `setProgram` is called for the ShaderMaterial.
- `refreshMaterial` is true because it's the first time rendering it, meaning the first block in this condition runs.
- The uniforms are updated but `uniformsNeedUpdate` remains true because the second block is not run even though the uniforms have already been updated.

**Frame 2**

- `setProgram` is called again for the ShaderMaterial.
- `refreshMaterial` is false because nothing else has changed.
- `uniformsNeedUpdate` is true because it was never set to false in frame 1 after updating the uniforms.
- The second `material.isShaderMaterial && material.uniformsNeedUpdate` block is run and needlessly updates the uniforms.

It doesn't update the uniforms twice in one from but it will needlessly update the uniforms in the next frame unless I'm missing something.

Reply:
Okay, I see now. We can basically do the following (it's actually the code from #15581. Unfortunately, the branch was deleted so I remembered incorrectly the code).
```
if ( material.isShaderMaterial ) material.uniformsNeedUpdate = false; // #15581
```
@amakaseev Can you please undo your change and put the following line at the end of the previous `if` block.